File: src/lib/yahria/policy-engine.ts
// ═══════════════════════════════════════════════════════════════
// YAHRIA KERNEL — Policy Control Plane (Domain 12)
// Doc ID: YAHRIA-KRN-004 | Invariants: INV-120, INV-121, INV-122, INV-123, INV-133
// Precedence: POLICY DENY > MODEL > AGENT > TOOL > LOCAL CONVENIENCE
// ═══════════════════════════════════════════════════════════════

import type { PolicyRequest, PolicyEvaluation, PolicyEffect } from './types';

export interface PolicyRuleDef {
  ruleId: string;
  name: string;
  effect: PolicyEffect;
  scope: string;           // TOOL | EXECUTION | FILESYSTEM | NETWORK | EVOLUTION | MODEL
  action: string;          // exact action or wildcard "fs.*"
  resource: string;        // exact resource or wildcard "*"
  priority: number;        // lower = evaluated first
  reason: string;
  version: string;
}

// Seed canonical policy set — DENY BY DEFAULT (INV-052, INV-133)
export const SEED_POLICY_RULES: PolicyRuleDef[] = [
  { ruleId: 'POL-001', name: 'Host secrets are never readable', effect: 'DENY', scope: 'FILESYSTEM', action: 'filesystem.read', resource: 'host.secrets', priority: 1, reason: 'INV-053: sandboxed execution MUST NOT access credentials or production secrets', version: '1.0.0' },
  { ruleId: 'POL-002', name: 'Host workspace never mounted RW', effect: 'DENY', scope: 'FILESYSTEM', action: 'filesystem.write', resource: 'host.workspace', priority: 1, reason: 'FS-001: host workspace never mounted RW into agent container', version: '1.0.0' },
  { ruleId: 'POL-003', name: 'No direct network egress from sandbox', effect: 'DENY', scope: 'NETWORK', action: 'network.egress', resource: '*', priority: 2, reason: 'INV-051: network isolation explicit; egress requires explicit policy', version: '1.0.0' },
  { ruleId: 'POL-004', name: 'Self-evolution cannot promote directly', effect: 'DENY', scope: 'EVOLUTION', action: 'evolution.promote', resource: 'production', priority: 1, reason: 'INV-160/161/162: D.8 governed by D.6.11 — experiment, benchmark, approve first', version: '1.0.0' },
  { ruleId: 'POL-005', name: 'Constitutional amendment is human-only', effect: 'DENY', scope: 'EVOLUTION', action: 'constitution.amend', resource: '*', priority: 1, reason: 'Global invariants evolve only through constitutional amendment', version: '1.0.0' },
  { ruleId: 'POL-006', name: 'Tool execution requires registration + authorization', effect: 'REQUIRE_APPROVAL', scope: 'TOOL', action: 'tool.execute', resource: 'unregistered.*', priority: 3, reason: 'INV-062: REGISTERED ≠ AUTHORIZED', version: '1.0.0' },
  { ruleId: 'POL-007', name: 'Sandboxed file writes allowed in overlay only', effect: 'ALLOW', scope: 'FILESYSTEM', action: 'filesystem.write', resource: 'workspace.overlay', priority: 10, reason: 'OverlayFS upper layer is the isolated RW surface', version: '1.0.0' },
  { ruleId: 'POL-008', name: 'Read-only inspection of workspace allowed', effect: 'ALLOW', scope: 'FILESYSTEM', action: 'filesystem.read', resource: 'workspace.snapshot', priority: 10, reason: 'Lower layer is read-only (FS-002)', version: '1.0.0' },
  { ruleId: 'POL-009', name: 'Bounded executions allowed in sandbox', effect: 'ALLOW', scope: 'EXECUTION', action: 'execution.run', resource: 'sandbox.*', priority: 10, reason: 'INV-042: resource bounded executions inside approved boundaries', version: '1.0.0' },
  { ruleId: 'POL-010', name: 'Model inference allowed with logging', effect: 'ALLOW', scope: 'MODEL', action: 'model.inference', resource: 'router.*', priority: 10, reason: 'INV-080: output is not fact; verdict requires verification', version: '1.0.0' },
];

// Final fallback — DENY BY DEFAULT
export const DEFAULT_EFFECT: PolicyEffect = 'DENY';
export const DEFAULT_REASON = 'INV-052/INV-133: no explicit policy matched → DEFAULT DENY. Security failure is not success.';

export function evaluatePolicy(request: PolicyRequest, rules: PolicyRuleDef[]): PolicyEvaluation {
  const sorted = [...rules].filter((r) => r.active !== false).sort((a, b) => a.priority - b.priority);
  for (const rule of sorted) {
    const actionMatch = rule.action.endsWith('*')
      ? request.action.startsWith(rule.action.slice(0, -1))
      : rule.action === request.action;
    const resourceMatch = rule.resource === '*' || rule.resource === request.resource ||
      (rule.resource.endsWith('.*') && request.resource.startsWith(rule.resource.slice(0, -1)));
    if (actionMatch && resourceMatch) {
      return {
        effect: rule.effect,
        matchedRule: rule.ruleId,
        reason: rule.reason,
        precedence: `POLICY ${rule.effect} > MODEL > AGENT > TOOL > LOCAL (rule ${rule.ruleId} v${rule.version}, priority ${rule.priority})`,
      };
    }
  }
  return {
    effect: DEFAULT_EFFECT,
    matchedRule: null,
    reason: DEFAULT_REASON,
    precedence: 'POLICY DENY (default) > MODEL > AGENT > TOOL > LOCAL',
  };
}

File: recovery/snapshot-full/yahria-core/kernel/policy_engine.py
# ═══════════════════════════════════════════════════════════════
# YAHRIA KERNEL (Python port) — Policy Control Plane (Domain 12)
# Doc ID: YAHRIA-KRN-004-PY | Invariants: INV-120, INV-121, INV-122, INV-123, INV-133
# Port of: src/lib/yahria/policy-engine.ts (YAHRIA-KRN-004)
# Precedence: POLICY DENY > MODEL > AGENT > TOOL > LOCAL CONVENIENCE
# ═══════════════════════════════════════════════════════════════

from dataclasses import dataclass
from typing import List, Optional

from kernel.types import PolicyEvaluation, PolicyRequest


@dataclass(frozen=True)
class PolicyRuleDef:
    ruleId: str
    name: str
    effect: str              # PolicyEffect value
    scope: str               # TOOL | EXECUTION | FILESYSTEM | NETWORK | EVOLUTION | MODEL
    action: str              # exact action or wildcard "fs.*"
    resource: str            # exact resource or wildcard "*"
    priority: int            # lower = evaluated first
    active: Optional[bool]   # optional enable switch (None = active)
    reason: str
    version: str

    def to_dict(self) -> dict:
        d = {
            "ruleId": self.ruleId, "name": self.name, "effect": self.effect,
            "scope": self.scope, "action": self.action, "resource": self.resource,
            "priority": self.priority, "reason": self.reason, "version": self.version,
        }
        if self.active is not None:
            d["active"] = self.active
        return d


# Seed canonical policy set — DENY BY DEFAULT (INV-052, INV-133)
SEED_POLICY_RULES: List[PolicyRuleDef] = [
    PolicyRuleDef("POL-001", "Host secrets are never readable", "DENY", "FILESYSTEM", "filesystem.read", "host.secrets", 1, None, "INV-053: sandboxed execution MUST NOT access credentials or production secrets", "1.0.0"),
    PolicyRuleDef("POL-002", "Host workspace never mounted RW", "DENY", "FILESYSTEM", "filesystem.write", "host.workspace", 1, None, "FS-001: host workspace never mounted RW into agent container", "1.0.0"),
    PolicyRuleDef("POL-003", "No direct network egress from sandbox", "DENY", "NETWORK", "network.egress", "*", 2, None, "INV-051: network isolation explicit; egress requires explicit policy", "1.0.0"),
    PolicyRuleDef("POL-004", "Self-evolution cannot promote directly", "DENY", "EVOLUTION", "evolution.promote", "production", 1, None, "INV-160/161/162: D.8 governed by D.6.11 — experiment, benchmark, approve first", "1.0.0"),
    PolicyRuleDef("POL-005", "Constitutional amendment is human-only", "DENY", "EVOLUTION", "constitution.amend", "*", 1, None, "Global invariants evolve only through constitutional amendment", "1.0.0"),
    PolicyRuleDef("POL-006", "Tool execution requires registration + authorization", "REQUIRE_APPROVAL", "TOOL", "tool.execute", "unregistered.*", 3, None, "INV-062: REGISTERED ≠ AUTHORIZED", "1.0.0"),
    PolicyRuleDef("POL-007", "Sandboxed file writes allowed in overlay only", "ALLOW", "FILESYSTEM", "filesystem.write", "workspace.overlay", 10, None, "OverlayFS upper layer is the isolated RW surface", "1.0.0"),
    PolicyRuleDef("POL-008", "Read-only inspection of workspace allowed", "ALLOW", "FILESYSTEM", "filesystem.read", "workspace.snapshot", 10, None, "Lower layer is read-only (FS-002)", "1.0.0"),
    PolicyRuleDef("POL-009", "Bounded executions allowed in sandbox", "ALLOW", "EXECUTION", "execution.run", "sandbox.*", 10, None, "INV-042: resource bounded executions inside approved boundaries", "1.0.0"),
    PolicyRuleDef("POL-010", "Model inference allowed with logging", "ALLOW", "MODEL", "model.inference", "router.*", 10, None, "INV-080: output is not fact; verdict requires verification", "1.0.0"),
]

# Final fallback — DENY BY DEFAULT
DEFAULT_EFFECT = "DENY"
DEFAULT_REASON = "INV-052/INV-133: no explicit policy matched → DEFAULT DENY. Security failure is not success."


def evaluate_policy(request: PolicyRequest, rules: List[PolicyRuleDef]) -> PolicyEvaluation:
    # Mirror of TS: filter active !== false, then stable sort by priority.
    active_rules = [r for r in rules if r.active is not False]
    sorted_rules = sorted(active_rules, key=lambda r: r.priority)

    for rule in sorted_rules:
        if rule.action.endswith("*"):
            action_match = request.action.startswith(rule.action[:-1])
        else:
            action_match = rule.action == request.action
        resource_match = (
            rule.resource == "*"
            or rule.resource == request.resource
            or (rule.resource.endswith(".*") and request.resource.startswith(rule.resource[:-1]))
        )
        if action_match and resource_match:
            return PolicyEvaluation(
                effect=rule.effect,
                matchedRule=rule.ruleId,
                reason=rule.reason,
                precedence=f"POLICY {rule.effect} > MODEL > AGENT > TOOL > LOCAL (rule {rule.ruleId} v{rule.version}, priority {rule.priority})",
            )

    return PolicyEvaluation(
        effect=DEFAULT_EFFECT,
        matchedRule=None,
        reason=DEFAULT_REASON,
        precedence="POLICY DENY (default) > MODEL > AGENT > TOOL > LOCAL",
    )
